### ISSUE TITLE
How to save/load the partial model for fine-tuning/transfer-learning in TF2.1?

### ISSUE THREAD

**Issue by hgffly** [tensorflow/tensorflow#37894]
I would like to build my own base model and train it with big dataset. After training, I save the base model. I have another customized model and I want to load the weights of first two layers from the base model. How should I achieve it in Tensorflow 2.1.0, thanks.

Sample codes:

    import os
    os.environ["CUDA_VISIBLE_DEVICES"]="" 
    import tensorflow as tf
    from tensorflow import keras
    from tensorflow.keras import layers

    class BaseModel():
        def __init__(self):
            inputs = keras.Input(shape=(32, 32, 3))
            x = inputs
            x = layers.Conv2D(32, 3, padding='same', activation=tf.nn.relu)(x)
            x = layers.MaxPool2D()(x)
            x = layers.Conv2D(64, 3, padding='same', activation=tf.nn.relu)(x)

            x = layers.Flatten()(x)

            x = layers.Dense(500, activation=tf.nn.relu)(x)

            outputs = layers.Dense(1000, activation=tf.nn.softmax)(x)

            self.model = keras.Model(inputs=inputs, outputs=outputs)

        def __call__(self, inputs):
            return self.model(inputs)

    bm = BaseModel()  # the model for pretraining
    bm.model.save_weights('base_model') # save the pretrained model

    class MyModel():
        def __init__(self):
            inputs = keras.Input(shape=(32, 32, 3))
            x = inputs
            x = layers.Conv2D(32, 3, padding='same', activation=tf.nn.relu)(x)
            x = layers.MaxPool2D()(x)
            x = layers.Conv2D(64, 3, padding='same', activation=tf.nn.relu)(x)
            
            x = layers.Conv2D(128, 3, padding='same', activation=tf.nn.relu)(x)

            x = layers.Flatten()(x)

            x = layers.Dense(1000, activation=tf.nn.relu)(x)

            outputs = layers.Dense(10, activation=tf.nn.softmax)(x)

            self.model = keras.Model(inputs=inputs, outputs=outputs)

        def __call__(self, inputs):
            return self.model(inputs)

    mm = MyModel()  # the model for my customized applications
    mm.model.load_weights('base_model')  # load the pretrained model with the first two conv layers

    # further fine-tuning or transfer learning

**Comment by jvishnuvardhan** [comment]
@hgffly I have used the custom model available in TF website demonstrate the idea. Using subclassed models is little different from Keras Sequential and functional model. I have used Subclassed model to demo the idea as follows. [Here](https://colab.research.google.com/gist/jvishnuvardhan/0af2d51513b5e17b7abd4431b0d7ddf7/loading_weights.ipynb is a colab gist for your reference. 

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

## base model
class ThreeLayerMLP(keras.Model):

  def __init__(self, name=None):
    super(ThreeLayerMLP, self).__init__(name=name)
    self.dense_1 = layers.Dense(64, activation='relu', name='dense_1')
    self.dense_2 = layers.Dense(64, activation='relu', name='dense_2')
    self.pred_layer = layers.Dense(10, name='predictions')

  def call(self, inputs):
    x = self.dense_1(inputs)
    x = self.dense_2(x)
    return self.pred_layer(x)

def get_model():
  return ThreeLayerMLP(name='3_layer_mlp')

base_model = get_model()

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784).astype('float32') / 255
x_test = x_test.reshape(10000, 784).astype('float32') / 255

base_model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              optimizer=keras.optimizers.RMSprop())
history = base_model.fit(x_train, y_train,
                    batch_size=64,
                    epochs=1)
#saving weights
base_model.save_weights('./base_model_weights', save_format='tf')

## custom model
class MyCustomModel(keras.Model):

  def __init__(self, name=None):
    super(MyCustomModel, self).__init__(name=name)
    self.dense_1 = layers.Dense(64, activation='relu', name='dense_1')
    self.dense_2 = layers.Dense(64, activation='relu', name='dense_2')
    self.pred_layer = layers.Dense(10, name='predictions')

  def call(self, inputs):
    x = self.dense_1(inputs)
    x = self.dense_2(x)
    return self.pred_layer(x)

def get_custom_model():
  return MyCustomModel(name='my_custom_model')

my_custom_model = get_custom_model()

my_custom_model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                  optimizer=keras.optimizers.RMSprop())

# This initializes the variables used by the optimizers,
# as well as any stateful metric variables
my_custom_model.train_on_batch(x_train[:1], y_train[:1])

# Load the state of the old model (to load weights for all layers)
# my_custom_model.load_weights('path_to_my_weights')

layer_dict = dict([(layer.name, layer) for layer in base_model.layers])
print(layer_dict)

# my_custom_model.trainable = True
# loading the weights from base_model
for layer in my_custom_model.layers:
  layer_name = layer.name
  #print(layer.name)
  layer.set_weights = layer_dict[layer_name].get_weights()
```

There are several ways to access and assign model weights. what I have shown in one of them.

Please close the issue. In future please post this kind of support questions in stackoverflow as GitHub is mainly for Bugs and performance related questions. Thanks!

**Comment by hgffly** [comment]
Thanks for help, but the codes should be modified as following:
layer.set_weights(layer_dict[layer_name].get_weights())

**Comment by yjiangling** [comment]
> @hgffly I have used the custom model available in TF website demonstrate the idea. Using subclassed models is little different from Keras Sequential and functional model. I have used Subclassed model to demo the idea as follows. [Here](https://colab.research.google.com/gist/jvishnuvardhan/0af2d51513b5e17b7abd4431b0d7ddf7/loading_weights.ipynb) is a colab gist for your reference.
> 
> ```
> import tensorflow as tf
> from tensorflow import keras
> from tensorflow.keras import layers
> 
> ## base model
> class ThreeLayerMLP(keras.Model):
> 
>   def __init__(self, name=None):
>     super(ThreeLayerMLP, self).__init__(name=name)
>     self.dense_1 = layers.Dense(64, activation='relu', name='dense_1')
>     self.dense_2 = layers.Dense(64, activation='relu', name='dense_2')
>     self.pred_layer = layers.Dense(10, name='predictions')
> 
>   def call(self, inputs):
>     x = self.dense_1(inputs)
>     x = self.dense_2(x)
>     return self.pred_layer(x)
> 
> def get_model():
>   return ThreeLayerMLP(name='3_layer_mlp')
> 
> base_model = get_model()
> 
> (x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
> x_train = x_train.reshape(60000, 784).astype('float32') / 255
> x_test = x_test.reshape(10000, 784).astype('float32') / 255
> 
> base_model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
>               optimizer=keras.optimizers.RMSprop())
> history = base_model.fit(x_train, y_train,
>                     batch_size=64,
>                     epochs=1)
> #saving weights
> base_model.save_weights('./base_model_weights', save_format='tf')
> 
> ## custom model
> class MyCustomModel(keras.Model):
> 
>   def __init__(self, name=None):
>     super(MyCustomModel, self).__init__(name=name)
>     self.dense_1 = layers.Dense(64, activation='relu', name='dense_1')
>     self.dense_2 = layers.Dense(64, activation='relu', name='dense_2')
>     self.pred_layer = layers.Dense(10, name='predictions')
> 
>   def call(self, inputs):
>     x = self.dense_1(inputs)
>     x = self.dense_2(x)
>     return self.pred_layer(x)
> 
> def get_custom_model():
>   return MyCustomModel(name='my_custom_model')
> 
> my_custom_model = get_custom_model()
> 
> my_custom_model.compile(loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
>                   optimizer=keras.optimizers.RMSprop())
> 
> # This initializes the variables used by the optimizers,
> # as well as any stateful metric variables
> my_custom_model.train_on_batch(x_train[:1], y_train[:1])
> 
> # Load the state of the old model (to load weights for all layers)
> # my_custom_model.load_weights('path_to_my_weights')
> 
> layer_dict = dict([(layer.name, layer) for layer in base_model.layers])
> print(layer_dict)
> 
> # my_custom_model.trainable = True
> # loading the weights from base_model
> for layer in my_custom_model.layers:
>   layer_name = layer.name
>   #print(layer.name)
>   layer.set_weights = layer_dict[layer_name].get_weights()
> ```
> 
> There are several ways to access and assign model weights. what I have shown in one of them.
> 
> Please close the issue. In future please post this kind of support questions in stackoverflow as GitHub is mainly for Bugs and performance related questions. Thanks!


For example, there is a saved checkpoint model trained with SourceModel, how to transfer the compatible parameters to the the TargetModel and start a new train?

	# 自定义层示例
	class CustomLayer(tf.keras.layers.Layer):
	    def __init__(self, unit1, unit2, **kwargs):
	        super(CustomLayer, self).__init__(**kwargs)
	        self.dense1 = tf.keras.layers.Dense(unit1, activation='relu')
	        self.dense2 = tf.keras.layers.Dense(unit2, activation='softmax')
	 
	    def call(self, inputs):
	        x = self.dense1(inputs)
	        return self.dense2(x)

	# 源模型示例
	class SourceModel(tf.keras.Model):
	    def __init__(self):
	        super(SourceModel, self).__init__()
	        self.layer = tf.keras.layers.Dense(128, activation='relu')
	        self.custom_layer = CustomLayer(64, 10)
	 
	    def call(self, inputs):
	        x = self.layer(inputs)
	        x = self.custom_layer(x)
	        return x

	# 目标模型示例
	class TargetModel(tf.keras.Model):
	    def __init__(self):
	        super(SourceModel, self).__init__()
	        self.layer = tf.keras.layers.Dense(128, activation='relu')
	        self.custom_layer = CustomLayer(64, 5)
	 
	    def call(self, inputs):
	        x = self.layer(inputs)
	        x = self.custom_layer(x)
	        return x

How to get the layer in the CustomLayer and transfer the weight to the TargetModel?
